modules/internal/ChapelDynamicLoading.chpl

```diff
@@ -0,0 +1,797 @@
+              if err {
+                on Locales[0] do errBuf[n] = err;
+                numErrs.fetchAdd(1);
+              } else if ptr {
+                var got = chpl_insertExternLocalPtrNoSync(ptr, idx);
+                assert(got == idx);
+              } else {
+                // TODO: Construct an error instead.
+                halt('Failed to fetch symbol!');
+              }
+            }
+          }
+
+          if numErrs.read() {
+            // If there were errors, then report the first error.
+            // TODO: Evict any pointers stored in the cache.
+            // TODO: Consolidate the reported errors.
+            for i in 0..<errBuf.size {
+              if errBuf[i] {
+                err = errBuf[i];
+                break;
+              }
+            }
+            // Clear the index since there was an error.
+            idx = 0;
+          }
+        }
+      }
+
+      // Fine if 'ret' is '0', that is the wide-index equivalent of 'nil'.
+      var ret = __primitive("cast", chpl_toExternProcType(t), idx);
+
+      return ret;
+    }
+  }
+
+  // This is a local buffer that only provides '[]' indexing and halts on an
+  // out-of-range index value. It halts instead of failing nicely because
+  // it should not fail (and perhaps we should disable the bounds checks).
+  record chpl_localBuffer {
+    type T;
+    var _ptr: c_ptr(T) = nil;
+    var _size: int = 0;
+
+    proc init(type T) {
+      this.T = T;
+      init this;
+    }
+
+    proc init(type T, size: int) {
+      this.T = T;
+      this._size = size;
+      init this;
+
+      if _size > 0 {
+        this._ptr = allocate(T, (size : uint(64)), true);
+        if _ptr == nil then halt('Out of memory!');
+      }
+    }
+
+    proc ref deinit() {
+      if _ptr != nil {
+        deallocate(_ptr);
+        _ptr = nil;
+      }
+    }
+
+    inline proc size do return _size;
+    inline proc ptr do return _ptr;
+    inline proc needsDestroy() param {
+      import MemMove;
+      return MemMove.needsDestroy(T);
+    }
+
+    inline proc this(idx: integral) ref: T {
+      if boundsChecking then
+        if idx < 0 || idx >= _size then
+          halt('Out of bounds!');
+      return _ptr[idx];
+    }
+
+    iter these() ref {
+      for idx in 0..<_size do yield _ptr[idx];
+    }
+  }
+
+  // This is a local hashtable that uses linear probing. I reinvent the wheel
+  // instead of using 'ChapelHashtable' so that the latter can have the
+  // luxury of using procedure pointers in its implementation (the code here
+  // is helping to implement procedure pointers, so it cannot itself use
+  // them without some difficulty).
+  //
+  // This hashtable does not use a tombstone field and expects key types to be
+  // simple scalars or POD types. It reserves a key with all 0-bits for use as
+  // the empty key. The value type may be any moveable type.
+  record chpl_localMap {
+    type K, V;
```

Comment:
I assumed this is mostly a clone of Map.map and did not dig too deeply into it

Reply:
It's not, it's a simpler map where the bitvalue `0` is the null key. Keys can only be scalar values. Additionally it may grow linearly as well. I've added a unit test as we talked about offline.